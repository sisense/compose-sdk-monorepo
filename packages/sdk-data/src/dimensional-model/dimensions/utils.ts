import { DataSource } from '../../interfaces.js';
import { createAttribute } from '../attributes.js';
import { createDimension, createDateDimension } from './dimensions.js';
import { Dimension, Attribute } from '../interfaces.js';
import { DataSourceField } from '../types.js';

/**
 * Function to convert data source fields to dimensions.
 * @param fields - The data source fields to convert.
 * @param dataSource - The data source title.
 * @returns - The dimensions created from the data source fields.
 * @internal
 */
export function getDimensionsFromDataSourceFields(
  fields: DataSourceField[],
  dataSource: DataSource,
): Dimension[] {
  const attributeEntries = fields.map((field) => createAttributeEntry(field, dataSource));

  const groupedConfigs = groupAttributesByDimension(attributeEntries);
  return Object.values(groupedConfigs).map((config) => createDimension(config));
}

/**
 * Given a DataSourceField and the data source title, creates an attribute entry
 * that includes the dimension name, attribute name, and the attribute itself.
 */
const createAttributeEntry = (
  field: DataSourceField,
  dataSource: DataSource,
): {
  dimension: DimensionDefinition;
  attribute: Attribute;
} => {
  const dimensionId = field.table;
  const dimensionName = field.tableTitle || dimensionId;
  const attributeId = field.id;
  const attributeName = field.title || attributeId;
  const dataSourceConfig = { title: dataSource, live: false };

  const attribute =
    field.dimtype === 'datetime'
      ? createDateDimension({
          name: attributeName,
          expression: attributeId,
          dataSource: dataSourceConfig,
        })
      : createAttribute({
          name: attributeName,
          type: field.dimtype === 'text' ? 'text-attribute' : 'numeric-attribute',
          expression: attributeId,
          dataSource: dataSourceConfig,
        });

  return {
    dimension: {
      id: dimensionId,
      name: dimensionName,
    },
    attribute,
  };
};

/**
 * Raw definition of a dimension taken from DataSourceField.
 */
type DimensionDefinition = {
  id: string;
  name: string;
};

type DimensionConfig = {
  expression: string;
  name: string;
} & Record<Exclude<string, 'name'>, Attribute>;

type DimensionConfigMap = Record<string, DimensionConfig>;
/**
 * Groups an array of attribute entries by their dimension name.
 * Returns an object whose keys are dimension names and values are configuration objects
 * that can be passed to createDimension.
 */
const groupAttributesByDimension = (
  entries: { dimension: DimensionDefinition; attribute: Attribute }[],
): DimensionConfigMap =>
  entries.reduce<DimensionConfigMap>((acc, { dimension, attribute }) => {
    const dimensionConfig: DimensionConfig = acc[dimension.name] || {
      name: dimension.name,
      expression: dimension.id,
    };
    return {
      ...acc,
      [dimension.name]: {
        ...dimensionConfig,
        [attribute.name]: attribute,
      },
    };
  }, {});
