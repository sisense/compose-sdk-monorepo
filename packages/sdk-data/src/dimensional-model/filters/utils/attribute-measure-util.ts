import {
  createAttributeHelper,
  createCalculatedMeasureHelper,
  createMeasureHelper,
  getGranularityFromJaql,
} from '../../../utils.js';
import { Attribute, BaseMeasure, CalculatedMeasure, LevelAttribute } from '../../interfaces.js';
import { FilterJaql } from '../../types.js';
import { CustomFormulaJaql, FilterJaqlInternal, RankingFilterJaql } from './types.js';

/**
 * Creates an attribute or level attribute from the provided filter JAQL object
 *
 * @param jaql - Filter JAQL object
 * @returns attribute or level attribute
 */
export const createAttributeFromFilterJaql = (
  jaql: FilterJaql | FilterJaqlInternal,
): Attribute | LevelAttribute => {
  const { dim, datatype, title, datasource: dataSource } = jaql;
  return createAttributeHelper({
    expression: dim,
    dataType: datatype,
    granularity: getGranularityFromJaql(jaql),
    title,
    dataSource,
  });
};

/**
 * Creates a measure from the provided filter JAQL object
 *
 * @param jaql - Filter JAQL object
 * @returns Measure
 */
export const createMeasureFromFilterJaql = (jaql: FilterJaqlInternal): BaseMeasure | undefined => {
  const { dim, title, datatype: dataType, agg, datasource: dataSource } = jaql;
  if (!agg) return undefined;
  return createMeasureHelper({
    expression: dim,
    dataType,
    agg,
    granularity: getGranularityFromJaql(jaql),
    format: undefined,
    sort: undefined,
    title,
    dataSource,
  });
};

/**
 * Creates a measure from the provided ranking filter JAQL object
 *
 * @param jaql - Ranking filter Jaql object
 * @returns Measure
 */
export const createMeasureFromRankingFilterJaql = (
  jaql: RankingFilterJaql | CustomFormulaJaql,
  rankingMessage?: string,
): BaseMeasure | CalculatedMeasure => {
  if ('formula' in jaql) return createCalculatedMeasureHelper(jaql);
  const { dim, datatype: dataType, agg } = jaql as RankingFilterJaql;
  return createMeasureHelper({
    expression: dim,
    granularity: getGranularityFromJaql(jaql),
    dataType,
    agg,
    format: undefined,
    sort: undefined,
    title: rankingMessage,
  });
};
