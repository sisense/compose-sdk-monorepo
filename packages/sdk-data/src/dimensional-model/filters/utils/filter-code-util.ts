/* eslint-disable @typescript-eslint/no-explicit-any */
import { Filter } from '../../interfaces.js';

/**
 * Stringifies the argument
 *
 * @param arg - argument to stringify
 * @returns stringified argument
 */
export function stringifyHelper(arg: any): string {
  try {
    if (arg === null || arg === undefined) {
      return JSON.stringify(arg);
    }

    if (arg === Object(arg) && 'composeCode' in arg) {
      return arg.composeCode;
    }

    if (Array.isArray(arg)) {
      return (
        '[' +
        arg
          .map((e) => {
            return stringifyHelper(e);
          })
          .join(', ') +
        ']'
      );
    }

    if (typeof arg === 'string') {
      return `'${arg}'`;
    }

    if (typeof arg === 'number' || !isNaN(arg)) {
      return arg;
    }
  } catch (e) {
    console.error(e);
  }

  return JSON.stringify(arg);
}

/**
 * High order function to construct compose code for filter factory functions
 *
 * @param func - filter factory function
 */
export function withComposeCode(func: (...args: any[]) => any): (...args: any[]) => any {
  return function (...args: any[]): any {
    const argValues = args.map(stringifyHelper).join(', ');
    const signature = `filterFactory.${func.name}(${argValues})`;
    const filter: Filter = func(...args);
    filter.composeCode = signature;
    return filter;
  };
}
