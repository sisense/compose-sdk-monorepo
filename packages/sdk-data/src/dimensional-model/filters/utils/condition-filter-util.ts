import { ConditionFilterJaql, ConditionFilterType } from './types.js';
import { Attribute, BaseMeasure, Filter, MembersFilterConfig } from '../../interfaces.js';
import { withComposeCode } from './filter-code-util.js';
import * as filterFactory from '../factory.js';
import { createMeasureFromRankingFilterJaql } from './attribute-measure-util.js';
import { TranslatableError } from '../../../translation/translatable-error.js';
import { simplifyFilterConfig } from '../filter-config-utils.js';

const isTopCondition = (filter: ConditionFilterJaql): boolean => filter.top !== undefined;
const isBottomCondition = (filter: ConditionFilterJaql): boolean => filter.bottom !== undefined;
const isExcludeCondition = (filter: ConditionFilterJaql): boolean => !!filter.exclude?.members;
const isMembersCondition = (filter: ConditionFilterJaql): boolean =>
  !!filter.members && !!filter.isCondition;
const isWithinCondition = (filter: ConditionFilterJaql): boolean =>
  filter.last?.anchor !== undefined || filter.next?.anchor !== undefined;
const isGreaterThanCondition = (filter: ConditionFilterJaql): boolean =>
  filter.fromNotEqual !== undefined;
const isGreaterThanOrEqualCondition = (filter: ConditionFilterJaql): boolean =>
  filter.from !== undefined && !filter.isBetween;
const isLessThanCondition = (filter: ConditionFilterJaql): boolean =>
  filter.toNotEqual !== undefined;
const isLessThanOrEqualCondition = (filter: ConditionFilterJaql): boolean =>
  filter.to !== undefined && !filter.isBetween;
const isEqualsCondition = (filter: ConditionFilterJaql): boolean => filter.equals !== undefined;
const isNotEqualCondition = (filter: ConditionFilterJaql): boolean =>
  filter.doesntEqual !== undefined;
const isEmptyCondition = (filter: ConditionFilterJaql): boolean =>
  !!(filter.equals === '' && filter.isEmpty);
const isNotEmptyCondition = (filter: ConditionFilterJaql): boolean =>
  !!(filter.doesntEqual === '' && filter.isEmpty);
const isContainsCondition = (filter: ConditionFilterJaql): boolean => filter.contains !== undefined;
const isDoesntContainCondition = (filter: ConditionFilterJaql): boolean =>
  filter.doesntContain !== undefined;
const isStartsWithCondition = (filter: ConditionFilterJaql): boolean =>
  filter.startsWith !== undefined;
const isDoesntStartsWithCondition = (filter: ConditionFilterJaql): boolean =>
  filter.doesntStartWith !== undefined;
const isEndsWithCondition = (filter: ConditionFilterJaql): boolean => filter.endsWith !== undefined;
const isDoesntEndWithCondition = (filter: ConditionFilterJaql): boolean =>
  filter.doesntEndWith !== undefined;
const isBetweenCondition = (filter: ConditionFilterJaql): boolean =>
  filter.from !== undefined && filter.to !== undefined;
const isNotBetweenCondition = (filter: ConditionFilterJaql): boolean =>
  filter.exclude?.from !== undefined && filter.exclude?.to !== undefined;
const isMultipleCondition = (filter: ConditionFilterJaql): boolean => !!(filter.or || filter.and);

export const getSelectedConditionOption = (filter: ConditionFilterJaql): ConditionFilterType => {
  if (isBottomCondition(filter)) return ConditionFilterType.BOTTOM;
  if (isTopCondition(filter)) return ConditionFilterType.TOP;
  if (isExcludeCondition(filter)) return ConditionFilterType.IS_NOT;
  if (isWithinCondition(filter)) return ConditionFilterType.IS_WITHIN;
  if (isGreaterThanCondition(filter)) return ConditionFilterType.GREATER_THAN;
  if (isLessThanCondition(filter)) return ConditionFilterType.LESS_THAN;
  if (isEqualsCondition(filter)) return ConditionFilterType.EQUALS;
  if (isNotEqualCondition(filter)) return ConditionFilterType.DOESNT_EQUAL;
  if (isEmptyCondition(filter)) return ConditionFilterType.IS_EMPTY;
  if (isNotEmptyCondition(filter)) return ConditionFilterType.IS_NOT_EMPTY;
  if (isContainsCondition(filter)) return ConditionFilterType.CONTAINS;
  if (isDoesntContainCondition(filter)) return ConditionFilterType.DOESNT_CONTAIN;
  if (isDoesntEndWithCondition(filter)) return ConditionFilterType.DOESNT_END_WITH;
  if (isDoesntStartsWithCondition(filter)) return ConditionFilterType.DOESNT_START_WITH;
  if (isEndsWithCondition(filter)) return ConditionFilterType.ENDS_WITH;
  if (isStartsWithCondition(filter)) return ConditionFilterType.STARTS_WITH;
  if (isNotBetweenCondition(filter)) return ConditionFilterType.IS_NOT_BETWEEN;
  if (isMembersCondition(filter)) return ConditionFilterType.IS;
  if (isMultipleCondition(filter)) return ConditionFilterType.MULTIPLE_CONDITION;
  // Need to verify BETWEEN case before the GREATER_THAN_OR_EQUAL and LESS_THAN_OR_EQUAL due to missing `filter.isBetween` property in some cases
  if (isBetweenCondition(filter)) return ConditionFilterType.BETWEEN;
  if (isGreaterThanOrEqualCondition(filter)) return ConditionFilterType.GREATER_THAN_OR_EQUAL;
  if (isLessThanOrEqualCondition(filter)) return ConditionFilterType.LESS_THAN_OR_EQUAL;

  return ConditionFilterType.NONE;
};

/**
 * Creates an attribute filter from the provided attribute and condition filter JAQL object
 *
 * @param attribute - Provided attribute
 * @param conditionFilterJaql - Condition filter JAQL object
 * @param guid - GUID for the filter
 * @returns attribute filter
 */
export const createAttributeFilterFromConditionFilterJaql = (
  attribute: Attribute,
  conditionFilterJaql: ConditionFilterJaql,
  guid: string,
): Filter => {
  const conditionType = getSelectedConditionOption(conditionFilterJaql);
  switch (conditionType) {
    case ConditionFilterType.BOTTOM:
      if (conditionFilterJaql.by) {
        return withComposeCode(filterFactory.bottomRanking, 'bottomRanking')(
          attribute,
          createMeasureFromRankingFilterJaql(
            conditionFilterJaql.by,
            conditionFilterJaql.rankingMessage,
          ),
          conditionFilterJaql[ConditionFilterType.BOTTOM] as number,
          { guid },
        );
      }
      break;
    case ConditionFilterType.EQUALS:
      return withComposeCode(filterFactory.equals, 'equals')(
        attribute,
        conditionFilterJaql[ConditionFilterType.EQUALS] as string | number,
        { guid },
      );
    case ConditionFilterType.DOESNT_EQUAL:
      return withComposeCode(filterFactory.doesntEqual, 'doesntEqual')(
        attribute,
        conditionFilterJaql[ConditionFilterType.DOESNT_EQUAL] as string | number,
        { guid },
      );

    case ConditionFilterType.GREATER_THAN:
      return withComposeCode(filterFactory.greaterThan, 'greaterThan')(
        attribute,
        conditionFilterJaql[ConditionFilterType.GREATER_THAN] as number,
        { guid },
      );
    case ConditionFilterType.GREATER_THAN_OR_EQUAL:
      return withComposeCode(filterFactory.greaterThanOrEqual, 'greaterThanOrEqual')(
        attribute,
        conditionFilterJaql[ConditionFilterType.GREATER_THAN_OR_EQUAL] as number,
        { guid },
      );
    case ConditionFilterType.TOP:
      if (conditionFilterJaql.by) {
        return withComposeCode(filterFactory.topRanking, 'topRanking')(
          attribute,
          createMeasureFromRankingFilterJaql(
            conditionFilterJaql.by,
            conditionFilterJaql.rankingMessage,
          ),
          conditionFilterJaql[ConditionFilterType.TOP] as number,
          { guid },
        );
      }
      break;
    case ConditionFilterType.STARTS_WITH:
      return withComposeCode(filterFactory.startsWith, 'startsWith')(
        attribute,
        conditionFilterJaql[ConditionFilterType.STARTS_WITH] as string,
        { guid },
      );
    case ConditionFilterType.DOESNT_START_WITH:
      return withComposeCode(filterFactory.doesntStartWith, 'doesntStartWith')(
        attribute,
        conditionFilterJaql[ConditionFilterType.DOESNT_START_WITH] as string,
        { guid },
      );
    case ConditionFilterType.ENDS_WITH:
      return withComposeCode(filterFactory.endsWith, 'endsWith')(
        attribute,
        conditionFilterJaql[ConditionFilterType.ENDS_WITH] as string,
        { guid },
      );
    case ConditionFilterType.DOESNT_END_WITH:
      return withComposeCode(filterFactory.doesntEndWith, 'doesntEndWith')(
        attribute,
        conditionFilterJaql[ConditionFilterType.DOESNT_END_WITH] as string,
        { guid },
      );
    case ConditionFilterType.CONTAINS:
      return withComposeCode(filterFactory.contains, 'contains')(
        attribute,
        conditionFilterJaql[ConditionFilterType.CONTAINS] as string,
        { guid },
      );
    case ConditionFilterType.DOESNT_CONTAIN:
      return withComposeCode(filterFactory.doesntContain, 'doesntContain')(
        attribute,
        conditionFilterJaql[ConditionFilterType.DOESNT_CONTAIN] as string,
        { guid },
      );
    case ConditionFilterType.LESS_THAN:
      return withComposeCode(filterFactory.lessThan, 'lessThan')(
        attribute,
        conditionFilterJaql[ConditionFilterType.LESS_THAN] as number,
        { guid },
      );
    case ConditionFilterType.LESS_THAN_OR_EQUAL:
      return withComposeCode(filterFactory.lessThanOrEqual, 'lessThanOrEqual')(
        attribute,
        conditionFilterJaql[ConditionFilterType.LESS_THAN_OR_EQUAL] as number,
        { guid },
      );
    case ConditionFilterType.BETWEEN:
      return withComposeCode(filterFactory.between, 'between')(
        attribute,
        conditionFilterJaql.from as number,
        conditionFilterJaql.to as number,
        { guid },
      );

    case ConditionFilterType.IS_NOT_BETWEEN:
      return withComposeCode(filterFactory.exclude, 'exclude')(
        withComposeCode(filterFactory.between, 'between')(
          attribute,
          conditionFilterJaql.exclude?.from as number,
          conditionFilterJaql.exclude?.to as number,
          { guid },
        ),
        undefined,
        { guid },
      );
    case ConditionFilterType.MULTIPLE_CONDITION:
      if (conditionFilterJaql.and) {
        return withComposeCode(filterFactory.intersection, 'intersection')(
          conditionFilterJaql.and.map((c) =>
            createAttributeFilterFromConditionFilterJaql(attribute, c, guid),
          ),
          { guid },
        );
      }
      if (conditionFilterJaql.or) {
        return withComposeCode(filterFactory.union, 'union')(
          conditionFilterJaql.or.map((c) =>
            createAttributeFilterFromConditionFilterJaql(attribute, c, guid),
          ),
          { guid },
        );
      }
      break;
    case ConditionFilterType.IS_NOT: {
      const deactivatedMembers =
        (conditionFilterJaql.filter?.turnedOff && conditionFilterJaql.filter.exclude?.members) ||
        [];
      const selectedMembers =
        conditionFilterJaql.exclude?.members?.filter(
          (member) => !deactivatedMembers.includes(member),
        ) || [];

      // use members filter with exclude instead of exclude filter
      const config: MembersFilterConfig = simplifyFilterConfig({
        guid,
        excludeMembers: true,
        enableMultiSelection: conditionFilterJaql.multiSelection ?? true,
        deactivatedMembers,
      });
      return withComposeCode(filterFactory.members, 'members')(attribute, selectedMembers, config);
    }
    case ConditionFilterType.AFTER:
    case ConditionFilterType.BEFORE:
    case ConditionFilterType.IS_EMPTY:
    case ConditionFilterType.IS_NOT_EMPTY:
      // TODO Handle these cases later; may need filterFactory function added first
      break;
  }

  throw new TranslatableError('errors.filter.unsupportedConditionFilter', {
    filter: JSON.stringify(conditionFilterJaql),
  });
};

/**
 * Creates a measure filter from the provided measure and condition filter JAQL object
 *
 * @param measure - Provided measure
 * @param conditionFilterJaql - Condition filter JAQL object
 * @param guid - GUID for the filter
 * @returns measure filter
 */
export const createMeasureFilterFromConditionFilterJaql = (
  measure: BaseMeasure,
  conditionFilterJaql: ConditionFilterJaql,
  guid: string,
): Filter => {
  const conditionType = getSelectedConditionOption(conditionFilterJaql);
  switch (conditionType) {
    case ConditionFilterType.EQUALS:
      return withComposeCode(filterFactory.measureEquals, 'measureEquals')(
        measure,
        conditionFilterJaql[ConditionFilterType.EQUALS] as number,
        { guid },
      );
    case ConditionFilterType.GREATER_THAN:
      return withComposeCode(filterFactory.measureGreaterThan, 'measureGreaterThan')(
        measure,
        conditionFilterJaql[ConditionFilterType.GREATER_THAN] as number,
        { guid },
      );
    case ConditionFilterType.GREATER_THAN_OR_EQUAL:
      return withComposeCode(filterFactory.measureGreaterThanOrEqual, 'measureGreaterThanOrEqual')(
        measure,
        conditionFilterJaql[ConditionFilterType.GREATER_THAN_OR_EQUAL] as number,
        { guid },
      );
    case ConditionFilterType.LESS_THAN:
      return withComposeCode(filterFactory.measureLessThan, 'measureLessThan')(
        measure,
        conditionFilterJaql[ConditionFilterType.LESS_THAN] as number,
        { guid },
      );
    case ConditionFilterType.LESS_THAN_OR_EQUAL:
      return withComposeCode(filterFactory.measureLessThanOrEqual, 'measureLessThanOrEqual')(
        measure,
        conditionFilterJaql[ConditionFilterType.LESS_THAN_OR_EQUAL] as number,
        { guid },
      );
    case ConditionFilterType.BETWEEN:
      return withComposeCode(filterFactory.measureBetween, 'measureBetween')(
        measure,
        conditionFilterJaql.from as number,
        conditionFilterJaql.to as number,
        { guid },
      );
  }
  throw new TranslatableError('errors.filter.unsupportedConditionFilter', {
    filter: JSON.stringify(conditionFilterJaql),
  });
};
