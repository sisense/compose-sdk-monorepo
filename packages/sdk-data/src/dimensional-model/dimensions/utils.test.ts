import { getDimensionsFromDataSourceFields } from './utils.js';
import { sampleEcommerceFields } from './__mocks__/sample-ecommerce-fields.js';

describe('getDimensionsFromDataSourceFields', () => {
  it('should return a list of dimensions', () => {
    const result = getDimensionsFromDataSourceFields(sampleEcommerceFields, 'Sample ECommerce');
    expect(result).toMatchSnapshot();
  });
});
