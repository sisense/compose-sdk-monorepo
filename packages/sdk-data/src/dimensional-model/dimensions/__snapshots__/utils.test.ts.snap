// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getDimensionsFromDataSourceFields > should return a list of dimensions 1`] = `
[
  {
    "__serializable": "DimensionalDimension",
    "attributes": [
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Brand.Brand",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Brand.Brand]",
        "name": "Brand",
        "type": "text-attribute",
      },
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Brand.BrandID",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Brand.Brand ID]",
        "name": "Brand ID",
        "type": "numeric-attribute",
      },
    ],
    "composeCode": "DM.ran.",
    "dataSource": undefined,
    "description": "",
    "dimensions": [],
    "expression": "Brand",
    "name": "Brand",
    "sort": 0,
    "type": "textdimension",
  },
  {
    "__serializable": "DimensionalDimension",
    "attributes": [
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Category.Category",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Category.Category]",
        "name": "Category",
        "type": "text-attribute",
      },
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Category.CategoryID",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Category.Category ID]",
        "name": "Category ID",
        "type": "numeric-attribute",
      },
    ],
    "composeCode": "DM.ategor.",
    "dataSource": undefined,
    "description": "",
    "dimensions": [],
    "expression": "Category",
    "name": "Category",
    "sort": 0,
    "type": "textdimension",
  },
  {
    "__serializable": "DimensionalDimension",
    "attributes": [
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Commerce.AgeRange",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Commerce.Age Range]",
        "name": "Age Range",
        "type": "text-attribute",
      },
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Commerce.BrandID",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Commerce.Brand ID]",
        "name": "Brand ID",
        "type": "numeric-attribute",
      },
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Commerce.CategoryID",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Commerce.Category ID]",
        "name": "Category ID",
        "type": "numeric-attribute",
      },
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Commerce.Condition",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Commerce.Condition]",
        "name": "Condition",
        "type": "text-attribute",
      },
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Commerce.Cost",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Commerce.Cost]",
        "name": "Cost",
        "type": "numeric-attribute",
      },
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Commerce.CountryID",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Commerce.Country ID]",
        "name": "Country ID",
        "type": "numeric-attribute",
      },
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Commerce.Gender",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Commerce.Gender]",
        "name": "Gender",
        "type": "text-attribute",
      },
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Commerce.Quantity",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Commerce.Quantity]",
        "name": "Quantity",
        "type": "numeric-attribute",
      },
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Commerce.Revenue",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Commerce.Revenue]",
        "name": "Revenue",
        "type": "numeric-attribute",
      },
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Commerce.VisitID",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Commerce.Visit ID]",
        "name": "Visit ID",
        "type": "numeric-attribute",
      },
    ],
    "composeCode": "DM.ommerc.",
    "dataSource": undefined,
    "description": "",
    "dimensions": [
      {
        "__serializable": "DimensionalDateDimension",
        "attributes": [
          {
            "__serializable": "DimensionalLevelAttribute",
            "composeCode": "DM.Commerce.Date.Years",
            "dataSource": {
              "live": false,
              "title": "Sample ECommerce",
            },
            "description": "",
            "expression": "[Commerce.Date (Calendar)]",
            "format": "yyyy",
            "granularity": "Years",
            "name": "Years",
            "type": "datelevel",
          },
          {
            "__serializable": "DimensionalLevelAttribute",
            "composeCode": "DM.Commerce.Date.Quarters",
            "dataSource": {
              "live": false,
              "title": "Sample ECommerce",
            },
            "description": "",
            "expression": "[Commerce.Date (Calendar)]",
            "format": "Q yyyy",
            "granularity": "Quarters",
            "name": "Quarters",
            "type": "datelevel",
          },
          {
            "__serializable": "DimensionalLevelAttribute",
            "composeCode": "DM.Commerce.Date.Months",
            "dataSource": {
              "live": false,
              "title": "Sample ECommerce",
            },
            "description": "",
            "expression": "[Commerce.Date (Calendar)]",
            "format": "yyyy-MM",
            "granularity": "Months",
            "name": "Months",
            "type": "datelevel",
          },
          {
            "__serializable": "DimensionalLevelAttribute",
            "composeCode": "DM.Commerce.Date.Weeks",
            "dataSource": {
              "live": false,
              "title": "Sample ECommerce",
            },
            "description": "",
            "expression": "[Commerce.Date (Calendar)]",
            "format": "ww yyyy",
            "granularity": "Weeks",
            "name": "Weeks",
            "type": "datelevel",
          },
          {
            "__serializable": "DimensionalLevelAttribute",
            "composeCode": "DM.Commerce.Date.Days",
            "dataSource": {
              "live": false,
              "title": "Sample ECommerce",
            },
            "description": "",
            "expression": "[Commerce.Date (Calendar)]",
            "format": "yyyy-MM-dd",
            "granularity": "Days",
            "name": "Days",
            "type": "datelevel",
          },
          {
            "__serializable": "DimensionalLevelAttribute",
            "composeCode": "DM.Commerce.Date.Hours",
            "dataSource": {
              "live": false,
              "title": "Sample ECommerce",
            },
            "description": "",
            "expression": "[Commerce.Date (Calendar)]",
            "format": "yyyy-MM-dd HH:mm",
            "granularity": "Hours",
            "name": "Hours",
            "type": "datelevel",
          },
          {
            "__serializable": "DimensionalLevelAttribute",
            "composeCode": "DM.Commerce.Date.MinutesRoundTo30",
            "dataSource": {
              "live": false,
              "title": "Sample ECommerce",
            },
            "description": "",
            "expression": "[Commerce.Date (Calendar)]",
            "format": "yyyy-MM-dd HH:mm",
            "granularity": "MinutesRoundTo30",
            "name": "MinutesRoundTo30",
            "type": "datelevel",
          },
          {
            "__serializable": "DimensionalLevelAttribute",
            "composeCode": "DM.Commerce.Date.MinutesRoundTo15",
            "dataSource": {
              "live": false,
              "title": "Sample ECommerce",
            },
            "description": "",
            "expression": "[Commerce.Date (Calendar)]",
            "format": "yyyy-MM-dd HH:mm",
            "granularity": "MinutesRoundTo15",
            "name": "MinutesRoundTo15",
            "type": "datelevel",
          },
          {
            "__serializable": "DimensionalLevelAttribute",
            "composeCode": "DM.Commerce.Date.Minutes",
            "dataSource": {
              "live": false,
              "title": "Sample ECommerce",
            },
            "description": "",
            "expression": "[Commerce.Date (Calendar)]",
            "format": "yyyy-MM-dd HH:mm",
            "granularity": "Minutes",
            "name": "Minutes",
            "type": "datelevel",
          },
          {
            "__serializable": "DimensionalLevelAttribute",
            "composeCode": "DM.Commerce.Date.Seconds",
            "dataSource": {
              "live": false,
              "title": "Sample ECommerce",
            },
            "description": "",
            "expression": "[Commerce.Date (Calendar)]",
            "format": "yyyy-MM-dd HH:mm:ss",
            "granularity": "Seconds",
            "name": "Seconds",
            "type": "datelevel",
          },
          {
            "__serializable": "DimensionalLevelAttribute",
            "composeCode": "DM.Commerce.Date.AggHours",
            "dataSource": {
              "live": false,
              "title": "Sample ECommerce",
            },
            "description": "",
            "expression": "[Commerce.Date (Calendar)]",
            "format": "HH:mm",
            "granularity": "AggHours",
            "name": "AggHours",
            "type": "datelevel",
          },
          {
            "__serializable": "DimensionalLevelAttribute",
            "composeCode": "DM.Commerce.Date.AggMinutesRoundTo30",
            "dataSource": {
              "live": false,
              "title": "Sample ECommerce",
            },
            "description": "",
            "expression": "[Commerce.Date (Calendar)]",
            "format": "HH:mm",
            "granularity": "AggMinutesRoundTo30",
            "name": "AggMinutesRoundTo30",
            "type": "datelevel",
          },
          {
            "__serializable": "DimensionalLevelAttribute",
            "composeCode": "DM.Commerce.Date.AggMinutesRoundTo15",
            "dataSource": {
              "live": false,
              "title": "Sample ECommerce",
            },
            "description": "",
            "expression": "[Commerce.Date (Calendar)]",
            "format": "HH:mm",
            "granularity": "AggMinutesRoundTo15",
            "name": "AggMinutesRoundTo15",
            "type": "datelevel",
          },
          {
            "__serializable": "DimensionalLevelAttribute",
            "composeCode": "DM.Commerce.Date.AggMinutesRoundTo1",
            "dataSource": {
              "live": false,
              "title": "Sample ECommerce",
            },
            "description": "",
            "expression": "[Commerce.Date (Calendar)]",
            "format": "HH:mm",
            "granularity": "AggMinutesRoundTo1",
            "name": "AggMinutesRoundTo1",
            "type": "datelevel",
          },
        ],
        "composeCode": "DM.Commerce.Date",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "dimensions": [],
        "expression": "[Commerce.Date (Calendar)]",
        "name": "Date",
        "sort": 0,
        "type": "datedimension",
      },
    ],
    "expression": "Commerce",
    "name": "Commerce",
    "sort": 0,
    "type": "textdimension",
  },
  {
    "__serializable": "DimensionalDimension",
    "attributes": [
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Country.Country",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Country.Country]",
        "name": "Country",
        "type": "text-attribute",
      },
      {
        "__serializable": "DimensionalAttribute",
        "composeCode": "DM.Country.CountryID",
        "dataSource": {
          "live": false,
          "title": "Sample ECommerce",
        },
        "description": "",
        "expression": "[Country.Country ID]",
        "name": "Country ID",
        "type": "numeric-attribute",
      },
    ],
    "composeCode": "DM.ountr.",
    "dataSource": undefined,
    "description": "",
    "dimensions": [],
    "expression": "Country",
    "name": "Country",
    "sort": 0,
    "type": "textdimension",
  },
]
`;
